src/pypi_name_mapping.rs

```diff
@@ -1,61 +1,100 @@
-use crate::config::get_cache_dir;
-use async_once_cell::OnceCell;
+use futures::{stream, StreamExt};
 use http_cache_reqwest::{CACacheManager, Cache, CacheMode, HttpCache, HttpCacheOptions};
 use miette::{IntoDiagnostic, WrapErr};
 use rattler_conda_types::{PackageUrl, RepoDataRecord};
-use reqwest::Client;
 use reqwest_middleware::ClientBuilder;
 use reqwest_retry::{policies::ExponentialBackoff, RetryTransientMiddleware};
-use serde::Deserialize;
+use serde::{Deserialize, Serialize};
 use std::{collections::HashMap, str::FromStr};
 use url::Url;
 
-#[derive(Deserialize)]
-struct CondaPyPiNameMapping {
+use crate::config::get_cache_dir;
+
+const STORAGE_URL: &str = "https://conda-mapping.prefix.dev";
+const HASH_DIR: &str = "hash-v0";
+
+#[derive(Debug, Deserialize, Serialize, Clone)]
+pub struct Package {
+    pypi_normalized_names: Option<Vec<String>>,
+    versions: Option<HashMap<String, pep440_rs::Version>>,
     conda_name: String,
-    pypi_name: String,
+    package_name: String,
+    direct_url: Option<Vec<String>>,
 }
 
 /// Downloads and caches the conda-forge conda-to-pypi name mapping.
-pub async fn conda_pypi_name_mapping() -> miette::Result<&'static HashMap<String, String>> {
-    static MAPPING: OnceCell<HashMap<String, String>> = OnceCell::new();
-    MAPPING.get_or_try_init(async {
-
-        // Construct a client with a retry policy and local caching
-        let retry_policy =
-            ExponentialBackoff::builder().build_with_max_retries(3);
-        let retry_strategy = RetryTransientMiddleware::new_with_policy(retry_policy);
-        let cache_strategy = Cache(HttpCache {
-            mode: CacheMode::Default,
-            manager: CACacheManager { path: get_cache_dir().expect("missing cache directory").join("http-cache") },
-            options: HttpCacheOptions::default(),
-        });
-        let client = ClientBuilder::new(Client::new())
-            .with(cache_strategy)
-            .with(retry_strategy)
-            .build();
-
-        let response = client.get("https://raw.githubusercontent.com/regro/cf-graph-countyfair/master/mappings/pypi/name_mapping.json").send().await
-            .into_diagnostic()
-            .context("failed to download pypi name mapping")?;
-        let mapping: Vec<CondaPyPiNameMapping> = response
-            .json()
-            .await
-            .into_diagnostic()
-            .context("failed to parse pypi name mapping")?;
-        let mapping_by_name: HashMap<_, _> = mapping
-            .into_iter()
-            .map(|m| (m.conda_name, m.pypi_name))
-            .collect();
-        Ok(mapping_by_name)
-    }).await
+pub async fn conda_pypi_name_mapping(
```

Comment:
Could this get a progress bar? At first I though pixi broke, but it was just downloading without telling me.